src/buildifier/buildifier.ts

```diff
@@ -0,0 +1,211 @@
+ */
+export async function buildifierLint(
+  fileContent: string,
+  type: BuildifierFileType,
+  lintMode: "warn",
+): Promise<BuildifierWarning[]>;
+
+export async function buildifierLint(
+  fileContent: string,
+  type: BuildifierFileType,
+  lintMode: BuildifierLintMode,
+): Promise<string | BuildifierWarning[]> {
+  const args = [`--type=${type}`, `--lint=${lintMode}`];
+  const outputs = await executeBuildifier(fileContent, args);
+  switch (lintMode) {
+    case "fix":
+      return outputs.stdout;
+    case "warn":
+      const trimmedOutput = outputs.stderr.trim();
+      if (trimmedOutput.length) {
+        return parseLintWarnings(trimmedOutput.split("\n"));
+      }
+      return [];
+  }
+}
+
+/**
+ * Returns the file type of a file with the given path.
+ *
+ * @param fsPath The file path, whose extension and basename are used to
+ *     determine the file type.
+ * @returns The buildifier type of the file.
+ */
+export function getBuildifierFileType(fsPath: string): BuildifierFileType {
+  // NOTE: The implementation here should be kept in sync with buildifier's
+  // automatic format detection (see:
+  // https://github.com/bazelbuild/buildtools/blob/d39e4d/build/lex.go#L88)
+  // so that user actions in the IDE are consistent with the behavior they
+  // would see running buildifier on the command line.
+  const parsedPath = path.parse(fsPath.toLowerCase());
+  if (parsedPath.ext === ".bzl" || parsedPath.ext === ".sky") {
+    return "bzl";
+  }
+  if (parsedPath.ext === ".build" || parsedPath.base === "build") {
+    return "build";
+  }
+  if (parsedPath.ext === ".workspace" || parsedPath.base === "workspace") {
+    return "workspace";
+  }
+  return "bzl";
+}
+
+/**
+ * Gets the path to the buildifier executable specified by the workspace
+ * configuration, if present.
+ *
+ * @returns The path to the buildifier executable specified in the workspace
+ *     configuration, or just "buildifier" if not present (in which case the
+ *     system path will be searched).
+ */
+export function getDefaultBuildifierExecutablePath(): string {
+  // Try to retrieve the executable from VS Code's settings. If it's not set,
+  // just use "buildifier" as the default and get it from the system PATH.
+  const bazelConfig = vscode.workspace.getConfiguration("bazel");
+  const buildifierExecutable = bazelConfig.buildifierExecutable as string;
+  if (buildifierExecutable.length === 0) {
+    return "buildifier";
+  }
+  return buildifierExecutable;
+}
+
+/**
+ * Parses the output of buildifier's {@code --lint=warn} mode and constructs
+ * objects representing the warnings.
+ *
+ * @param lines The lines of output from standard error.
+ */
+function parseLintWarnings(lines: string[]): BuildifierWarning[] {
+  const warnings = new Array<BuildifierWarning>();
+  let lineNumber = 0;
+  let category = "";
+  let message = "";
+
+  for (const line of lines) {
+    if (line.startsWith("stdin:")) {
```

Comment:
Some times it can be useful to include some example text of what you are parsing so someone that goes to change it in the future can see some examples of what it was trying to handle.  The multiline cases is likely the one to especially call out.

Reply:
Done.